test/extensions/filters/http/gcp_authn/gcp_authn_filter_integration_test.cc

```diff
@@ -217,14 +223,34 @@ TEST_P(GcpAuthnFilterIntegrationTest, BasicflowWithoutAudience) {
                                                          std::chrono::milliseconds(200)));
 
   // Send the request to cluster `cluster_0` and validate the response.
-  sendRequestToFirstClusterAndValidateResponse(/*with_audience=*/false);
+  sendRequestToDestinationAndValidateResponse(/*with_audience=*/false);
 
   // Verify request has been routed to `cluster_0` but not `gcp_authn` cluster.
   EXPECT_GE(test_server_->counter("cluster.gcp_authn.upstream_cx_total")->value(), 0);
   EXPECT_GE(test_server_->counter("cluster.cluster_0.upstream_cx_total")->value(), 1);
 
-  // Perform the clean-up.
-  cleanupUpstreamAndDownstream();
+  // Clean up the codec and connections.
+  cleanup();
+}
+
+// This test is sending the request with body to verify that the filter chain iteration which has
+// been stopped by `decodeHeader`'s return status will not be resumed by `decodeData`.
+TEST_P(GcpAuthnFilterIntegrationTest, SendRequestWithBody) {
+  initializeConfig(/*add_audience=*/true);
+  HttpIntegrationTest::initialize();
+  initiateClientConnection(/*send_request_body=*/true);
+  // Send the request with long wait time to intentionally delay the response from `gcp_authn`
+  // cluster.
+  EXPECT_TRUE(fake_upstreams_[1]->waitForHttpConnection(*dispatcher_, fake_gcp_authn_connection_,
+                                                        std::chrono::milliseconds(500000)));
+  // Send the request to `cluster_0` cluster.
+  AssertionResult assert_result = fake_upstreams_[0]->waitForHttpConnection(
+      *dispatcher_, fake_upstream_connection_, std::chrono::milliseconds(1));
```

Comment:
I think 1 millisecond is too short. Making it at least 1 second is more convincing that the request is not going through.

Reply:
Yep.  Done

Also I have verified this test will fail without return status change and will pass consistently with this PR